supplier/models.py

```diff
@@ -0,0 +1,104 @@
+from django.core.exceptions import FieldDoesNotExist, ObjectDoesNotExist, ValidationError
+
+
+class User(models.Model):
+    """An abstract class which holds basic user data:
+
+    Attributes:
+        user_name: a string which holds the user name, Also PK.
+        first_name: a string which holds the user first name.
+        last_name: a string which holds the user last name.
+        password: a string which holds the user password.
+    """
+
+    user_name = models.CharField(max_length=32,
+                                 primary_key=True,
+                                 null=False,
+                                 validators=[MinLengthValidator(6)],
+                                 help_text="User Name")
+    first_name = models.CharField(max_length=32,
+                                  validators=[MinLengthValidator(1)],
+                                  help_text="First Name")
+    last_name = models.CharField(max_length=32,
+                                 validators=[MinLengthValidator(1)],
+                                 help_text="Last Name")
+    password = models.CharField(max_length=16,
+                                validators=[MinLengthValidator(6)],
+                                null=False,
+                                help_text="Password")
+
+    class Meta:
+        abstract = True
+
+
+class Supplier(User):
+    """Supplier class inherit from User abstract class.
+
+    Supplier class lets the app user to do variety of actions
+    which evolves the Product and Delivery class.
+
+    Attributes:
+        business_name: a string which holds the Supplier business name.
+    """
+    business_name = models.CharField(max_length=32,
+                                     validators=[MinLengthValidator(6)],
+                                     null=False,
+                                     help_text="Business Name")
+
+    @staticmethod
+    def filter_by_field(field, field_value):
+        """Fetches list of Suppliers filtered with given field and given value.
+
+        Args:
+        field:
+            Supplier attribute.
+        field_value:
+            value which will filter the suppliers.
+
+        Returns:
+            returns a queryset containing the objects that match the specified filter conditions.
+
+        Raises:
+        FieldDoesNotExist error: if Supplier doesn't have the specified attribute.
+        """
+        if hasattr(Supplier, str(field)):
+            return Supplier.objects.filter(**{field: field_value})
+        raise FieldDoesNotExist
+
+    @staticmethod
+    def delete_supplier(supplier):
+        """Deletes supplier 'safely'
+
+        Args:
+            supplier - the Supplier wanted to be deleted
+
+        Returns:
+            None.
+
+        raises:
+        ObjectDoesNotExists error: if the supplier is not in DB.
+        """
+        if supplier not in Supplier.objects.all():
+            raise ObjectDoesNotExist
```

Comment:
Its supplierDoesNotExist and not objectdoesnotexist
better way to do that
just use the get method on the supplier table to get supplier.username since it is the primary key and if not exist suppleirdoesnotexist raised

Reply:
care to explain whats the difference? in both ways the object exists

Reply:
also what do you mean by supplierdoesnotexists? creating a new error class? doesnt seem that django auto generate class error...

Reply:
No,its like you have for example exception X and exception Y, Y extends X

if you try to catch Y you can catch it by catch X because it extends X

Now you are raising a universal exception objectdoesnotexist but you need SupplierDoesNotExist

in order to do that 
```suggestion
            Supplier.objects.get(user_name=supplier.user_name)
```
in this code the get method that is already their either gets you the supplier if it exists or it raises the exception SupplierDoesNotExist

Reply:
Since you are creating new model on django , there will be created new things along it like methods .... and exceptions.